### Rule of sum
    If there are n objects of the first type and there are k objects of the second type, then there are n + k objects of one of two types

In [1]:
print(['Alice', 'Bob', 'Charlie']  + [0, 1, 2, 3, 4])

['Alice', 'Bob', 'Charlie', 0, 1, 2, 3, 4]


### Rule of product
    If there are n objects of the first type and there are k objects of the second type, then there are n × k pairs of objects, the first of the first type and the second of the second type

In [3]:
from itertools import product
for i in product(['a', 'b', 'c'],['x', 'y', 'z']):
    print(''.join(i))

ax
ay
az
bx
by
bz
cx
cy
cz


In [6]:
# how many three digits palindronic number are there
# there are 9 posiblity for the first one and also for the last and 10 posiblity for the middle one
count = 0
for i in range(100, 1000):
    if str(i) == str(i)[::-1]:
        count += 1
#         print(i)
count

90

### tuples 
    The number of sequences of length k composed out of n symbols is n**k.
    with repitation

In [6]:
from itertools import product
for i in product(['a', 'b'], repeat=3): # with repeatation
    print(''.join(i))

aaa
aab
aba
abb
baa
bab
bba
bbb


#### number of path link
http://dm.compsciclub.ru/app/quiz-number-of-paths

### k permutations
    The number of sequences of length k with no repetitions composed out of n symbols is n(n − 1)· · ·(n − k + 1) = n!/(n−k)! .
    

In [1]:
from itertools import permutations
count = 0
for i in permutations('abcd', 4):
    count += 1
    print(''.join(i))
print()
print(count)

abcd
abdc
acbd
acdb
adbc
adcb
bacd
badc
bcad
bcda
bdac
bdca
cabd
cadb
cbad
cbda
cdab
cdba
dabc
dacb
dbac
dbca
dcab
dcba

24


## Tournament
    Five teams played a tournament: each team played with each other. What was the number of games?
    Each team played four games (with each ofthe other four teams)
        • Hence, there were 5 × 4 games
        • Do you see a flaw in this argument?

### let's arrange the game
#### ab ac ad ae bc bd be cd ce de
#### ba ca da ea cb db eb dc ec ed

     the flaw is each game is counted twice
     thus the number of game is (5*4)/2

### Theorem

    The number of games in a tournament with n teams (each pair of teams played each other exactly once) is n(n − 1)/2.

### proof

    • There are n choices of the first team in
    a game and (n − 1) choices of the second
    team
    • Each game is counted twice: the game
    between teams i and j is counted as ij and
    as ji
    • Thus, the total number of games is
    n(n − 1)/2

In [14]:
# the correct way of counting games played by each team
# mathmatical: 5 * 4/2=10
# in combination ab and ba is considered as same pair 
from itertools import combinations
count = 0
for i in combinations('abcde', 2):
    count += 1
    print(''.join(i))
print('total count is: ',count)

ab
ac
ad
ae
bc
bd
be
cd
ce
de
total count is:  10


In [17]:
s = 'abcde'
for i in range(len(s)):
    for k in range(i+1, len(s)):
        print(s[i], s[k])

a b
a c
a d
a e
b c
b d
b e
c d
c e
d e


### Let’s count the number of games recursively

    • Denote by T(n) the number of games in a tournament with n teams
    • There are two types of games:
        1. games that involve the first team: n − 1
        2. games that don’t involve it: T(n − 1)
     Hence, T(n) = (n − 1) + T(n − 1)

In [19]:
def T(n):
    if n <= 1:
        return 0
    else:
        return (n-1) + T(n-1)
T(5)

10

### Combination formula
   #### nCk = n!/k!(n-1)!
    five teams played with each other, so what is the number of games?
    5C2 = (5*4*3*2*1)/2*1*(3*2*1)
        = 10 games
    

### Pascal Triangle

    
    

In [4]:
def nCk(x, y): # C ( [x, y] ) i s equal to n choose k
    C = dict() 
    for row in range(x+1):
        C[row, 0] = 1
        C[row, row] = 1 # second row is considered as column 
        for col in range(1, row):
#             print(col,row)
            C [row, col] = C[row - 1, col - 1] + C[row-1, col]
    return (C[x, y])
nCk(5,2)

10

In [3]:
# hackerrank nCr table
# print the nth row of pascal traingle
# the formula is: nCr = nCr-1*(n-r+1)/r where 1<=r<=N
def solve(n):
    prev = 1
    print(prev, end='')
    for r in range(1, n+1):
        # prev = nCr-1
        # if the number is large, just print the last 9 digit
        current = (prev*(n - r + 1) // r)
        print(',',current%10**9, end='')
        prev=current
    print()
if __name__=='__main__':
    t = int(input())
    test = []
    for _ in range(t):
        test.append(int(input()))
    for n in test:
        solve(n)
# time complexity is O(n)
# auxillary space O(1)
# proof: the loop will run from 1 to n+1
        

2
5
10
1, 5, 10, 10, 5, 1
1, 10, 45, 120, 210, 252, 210, 120, 45, 10, 1


## Practice counting

In [4]:
# what is the number of five cards hand dealt off of a 52-deck card?
# (52C5)
nCk(52, 5)

2598960

In [5]:
# what is the number of ways selecting 5 cards with two hearts and 3 spades from 52 deck card
a = nCk(13, 2)
b = nCk(13, 3)
a * b

22308

In [8]:
# what is the number of non negative integers with at most four digits at least one of which is equal to 7?
all_possible_4_digits = 10**4
except_7 = 9**4
all_digits_with_7 = all_possible_4_digits - except_7
all_digits_with_7

3439

In [12]:
from itertools import product
count = 0
for t in product((range(10)), repeat=4):
    if 7 in t:
        count += 1
print(count)

3439


In [13]:
# What is the number of non-negative integers with at most four digits whose digits are increasing
# this is just choosing 4 elements out of 10
nCk(10,4)

210

In [14]:
from itertools import product
count = 0
for tup in product((range(10)), repeat=4):
    if tup[0] < tup[1] and tup[1] < tup[2] and tup[2] < tup[3]:
        count +=1 
print(count)

210


In [16]:
# What is the number of bit-strings (that is, strings consisting of 0's and 1's) of length 6 
# where the number of 0's is equal to the number of 1's?
from itertools import product
count = 0
for t in product(range(2), repeat=6):
    if list(t).count(0) == list(t).count(1):
        count += 1
print(count)

20


In [17]:
# What is the number of sequences of six digits where the number of even digits is equal to the number of odd digits?
from itertools import product
count = 0
for i in product(range(10), repeat=6):
    a = list(i)
    even = []
    odd = []
    for i in a:
        if i % 2 == 0:
            even.append(i)
        else:
            odd.append(i)
    if len(even) == len(odd):
        count += 1
print(count)
# this is slow

312500


## Week 3

In [29]:
# selecting two different people out of 3
# no item repeat
# we can also write as 3C2
from itertools import combinations
for i in combinations('abc', 2):
    print(i)

('a', 'b')
('a', 'c')
('b', 'c')


In [22]:
# selecting two different people out of 3
nCk(3, 2)

3

### salads
        We have an unlimited supply of tomatoes, bell peppers and lettuce. We want to make a salad out of 4 units among these           three ingredients (we do not have to use all ingredients). The order in which we use the ingredients does not matter.           How many different salads we can make?
        solution: a b c c [since we have to make a salad of 4 unit]
        let's keep delimiter to diferrentiate
        a d  b d c c
        we can convert it as:
        O O O O O O 
        We just have to choose where to put delimiter 
        so it's 6 choose 2
        
        The formula: unordered with repitation
        (k+n-1)C(n-1)
    

In [3]:
# order doesnt matter , it's still the same salad
# we can also think as: 4 voters are voting for one of three candidates 
# candidates can be chosen several times, as four voters can vote also the same person 
from itertools import combinations_with_replacement
count = 0
for i in combinations_with_replacement('abc', 4):
    print(i)
    count+=1
print(count)

('a', 'a', 'a', 'a')
('a', 'a', 'a', 'b')
('a', 'a', 'a', 'c')
('a', 'a', 'b', 'b')
('a', 'a', 'b', 'c')
('a', 'a', 'c', 'c')
('a', 'b', 'b', 'b')
('a', 'b', 'b', 'c')
('a', 'b', 'c', 'c')
('a', 'c', 'c', 'c')
('b', 'b', 'b', 'b')
('b', 'b', 'b', 'c')
('b', 'b', 'c', 'c')
('b', 'c', 'c', 'c')
('c', 'c', 'c', 'c')
15


In [2]:
nCk(6, 2)
# nCk(11, 2)

15

### Combinations with repitation
    The number of combination of size k  of n objects with repitation is equal to (k+n -1)C(n-1)

In [43]:
from itertools import combinations_with_replacement
count = 0
for i in combinations_with_replacement('abcde', 20):
#     print(i)
    count+=1
print(count)

10626


In [4]:
nCk(9, 4)

126

### Suppose there are 4 people and 9 different assignments. Each person should receive one assignment. Assignments for different people should be different. How many ways are there to do it?

In [3]:
from itertools import permutations
count = 0
for a in permutations("123456789", 4):
    count +=1
#     print("".join(a))
print(count)

3024


#### There are 4 people and 9 different assignments. We need to distribute all assignments among people. No assignment should be assigned to two people. Every person can be given arbitrary number of assignments from 0 to 9. How many ways are there to do it?

In [16]:
4**9

262144

### Distributing candies among people 

In [20]:
# There are 15 identical candies. How many ways are there to distribute them among 7 kids?
nCk(21, 6)

54264

In [24]:
# There are 15 identical candies. How many ways are there to distribute them among 7 kids in such 
#a way that each kid receives at least 1 candy?

We_are_left_with = 15- 7
nCk(14, 6)

3003

In [6]:
## There are six identical candies. How many ways are there to distibute them among 3 kids?
from itertools import combinations_with_replacement
count = 0
for i in combinations_with_replacement('123456', 3):
    count += 1
#     print(i)
print(count)

56


## Numbers with fixed sum of digits
## Stars and bars
    
    Q_1: How many non negative integers are there below 10000 such that their sum of digit is equal to 9
         a+b+c+d=12 where a, b, c, d are non negative int
         We split the 12 amongn them
         **|**|***|** >> 2 + 2 + 3 + 2 =9
         This is like choosing (9+4-1)C(4-1) which is 202
    Q_2: How many non negetive integers are there below 10000 such that their sum of digit is equal to 10
        By rule: (10+4-1)C(4-1) which is 286
        but the correct anser is 282 which is proved in code_for_Q_2 below
        with our approach, we can assign all 10 star in one position but this is not allowed in digit since digit range from
        0-9.so we have to subtract one star in total 4 from total combinations
        Thus, 286-4=282
    Q_3: How many 5 digits can be formed using digits 0-3 such that digits in number in non decreasing order
        ex:01233, 02233
        A0 + A1 + A2 + A3 = 5
        1 + 1 + 1 + 2 = 5 (01233)
        1 + 0 + 2+ 2 = 5 (02233)
        so this is like choosing 
        stars = 5 = number of digits = n
        bars = 4-1 = the available digits = k
        thus we get, (5+3-1)C(3-1)
## Numbers with non increasing digits
      Q_1: How many four-digit numbers are there such that their digits are non-increasing, that is each next digit is not           greater than the previous one? Three-digit numbers are also four-digit, they just start with 0.
      * * * *
      if we have picked 5323 from 0-9
      we have to place them like
      * * * *
      5 3 3 2 (non increasing order)
      so first we have to pick four digits, order of pick doesn't matter
      they can be order in unique way to place them in non increasing order
      We are dealing with combinations of size for out of 10 options with repitation
   
      n = 4 
      k = 10 
      (10+4-1)C(10-1) = 715


    
## Distributing Candies
    Q_1: In how many ways we can distribute 15 identical candi among 7 kids. 
    **|***|**||**|**|**** 2 + 3 + 2 + 0 + 2 + 2+ 4 = 15
    This is like choosing 3 places from 15 places
    (15 + 7 - 1)C(7-1) or 21C6= 54264

286

In [22]:
# code for Q_2
from itertools import product
count = 0
for i in product(range(10), repeat=4):
    if sum(i)==10:
#         print(i)
        count += 1
print(count)

282


In [40]:
nCk(2, 1 )

2

In [38]:
from itertools import combinations
count = 0
for i in combinations(range(7), 3):
    count += 1
print(count)

35


Alice has 7 textbooks and Bob has 5 textbooks. All textbooks are different. They would like to exchange three books each between each other for a weekend. That is, Alice gives Bob three of her books and Bob gives Alice three of his books. How many ways do they have to do it?

In [42]:

from itertools import combinations
count = 0
for a in combinations(["A1 ","A2 ","A3 ","A4 ","A5 ","A6 ","A7 "], 3):
    for b in combinations(["B1 ","B2 ","B3 ","B4 ","B5 "], 3):
        count += 1
        c=a+b
#         print(c)
print(count)

350


In [10]:
from itertools import combinations
count = 0
for i in combinations([(1, 2),(3,4),(5,6),(7,8),(9,0)], 3):
    print(i)
    count += 1
print(count)

((1, 2), (3, 4), (5, 6))
((1, 2), (3, 4), (7, 8))
((1, 2), (3, 4), (9, 0))
((1, 2), (5, 6), (7, 8))
((1, 2), (5, 6), (9, 0))
((1, 2), (7, 8), (9, 0))
((3, 4), (5, 6), (7, 8))
((3, 4), (5, 6), (9, 0))
((3, 4), (7, 8), (9, 0))
((5, 6), (7, 8), (9, 0))
10


In [23]:
from itertools import combinations
s ='abb'
count = 0
for i in range(1, len(s)+1):
    for k in combinations(s, i):
        if 'b' in k:
            
            print(k)
print(count)

('b',)
('b',)
('a', 'b')
('a', 'b')
('b', 'b')
('a', 'b', 'b')
0


In [28]:
# a", "b", "b", "ab", "bb", "abb". The substrings that contain 'b' are "b", "b", "ab", "bb", "abb".
for i in combinations(s, 2):
    print(i)

('a', 'b')
('a', 'b')
('b', 'b')


In [23]:
# hacker k candy shop
#  # the formula is: nCr = nCr-1*(n-r+1)/r where 1<=r<=N
def nth_line_pascal(n):
    prev = 1
    res = []
    res.append(prev)
    for r in range(1, n+1):
        current = prev * (n-r+1)//r
        res.append(current)
        prev = current
    return res
def solve(n, k):
    n_prime = n + k -1
    k_prime = n -1
    result = nth_line_pascal(n_prime)[k_prime]
    return result
if __name__ == '__main__':
    test = int(input())
    result = []
    for _ in range(test):
        n = int(input())
        k = int(input())
        output = solve(n, k)
        result.append(output%10**9)
    print(*result, sep='\n')

2
4
1
2
3
4
4


In [8]:
nth_line_pascal(4)

[1, 4, 6, 4, 1]

In [17]:
solve(2, 3)

4

In [13]:
a = [2, 3, 4]
a[1]

3